src/mysql.erl

```diff
@@ -175,32 +181,86 @@ start_link(Options) ->
 %%
-%% If the 3rd argument is a list, it executes a parameterized query. This is
-%% equivallent to query/4 with the query timeout as given to start_link/1.
+%% If the 3rd argument is a list, it executes a parameterized query and applies
+%% no filtering/mapping of the result rows. This is equivallent to query/5
+%% with undefined as filter/map function and the query timeout as given to
+%% start_link/1.
+%%
+%% If the 3rd argumeent is a function, it is used to filter/map the result
+%% rows. This is equivalent to query/4 with the query timeout given to
+%% start_link/1.
 %%
 %% If the 3rd argument is a timeout, it executes a plain query with this
-%% timeout.
+%% timeout and applies no filtering/mapping of the resultset rows.
 %%
 %% The return value is the same as for query/2.
 %%
 %% @see query/2.
 %% @see query/4.
--spec query(Conn, Query, Params | Timeout) -> Result
+-spec query(Conn, Query, Params | FilterMap | Timeout) -> Result
     when Conn :: connection(),
          Query :: iodata(),
          Timeout :: timeout(),
          Params :: [term()],
+         FilterMap :: query_filtermap(),
          Result :: query_result().
 query(Conn, Query, Params) when is_list(Params) ->
-    query_call(Conn, {param_query, Query, Params});
+    query_call(Conn, {param_query, Query, Params, undefined, undefined});
+query(Conn, Query, FilterMap) when FilterMap=:=undefined;
+        is_function(FilterMap, 1); is_function(FilterMap, 2) ->
+    query_call(Conn, {query, Query, FilterMap, undefined});
 query(Conn, Query, Timeout) when is_integer(Timeout); Timeout == infinity ->
-    query_call(Conn, {query, Query, Timeout}).
+    query_call(Conn, {query, Query, undefined, Timeout}).
 
-%% @doc Executes a parameterized query with a timeout.
+%% @doc Depending on the 3rd and 4th arguments this function does different
+%% things.
+%%
+%% If either the 3rd argument is a list, a prepared statement is created,
+%% executed and then cached for a certain time. If the same query is executed
+%% again when it is already cached, it does not need to be prepared again.
+%% The minimum time the prepared statement is cached can be specified using the
+%% option `{query_cache_time, Milliseconds}' to start_link/1.
+%% Conversely, if the 3rd argument is not a list, a plain query is executed.
+%%
+%% If the 4th argument is a timeout, the query is executed with that timeout.
+%% Conversely, if the 4th argument is not a timeout, the timeout given in
+%% start_link/1 is used.
+%%
+%% If either the 3rd or 4th argument is a function instead of a list or
+%% timeout, respectively, it is used to filter/map the result rows.
+%%
+%% The return value is the same as for query/2.
+-spec query(Conn, Query, Params, Timeout) -> Result
+        when Conn :: connection(),
+             Query :: iodata(),
+             Timeout :: timeout(),
+             Params :: [term()],
+             Result :: query_result();
+    (Conn, Query, FilterMap, Timeout) -> Result
```

Comment:
Thanks :)

Reply:
I just noticed that this does not work with edoc... edoc will only generate documentation for the first clause :\

Reply:
Bummer... the guys on #erlang say that's somewhat normal as edoc just doesn't seem to support the full extend of what `-spec`s offer :( I opened an issue at bugs.erlang.org for this, but I don't think much will come of it anytime soon as it's not a pressing issue.

Reply:
As it stands, I see no alternative other than weakening the spec to:
```erlang
-spec query(Conn, Query, Params | FilterMap, Timeout | FilterMap) -> Result
when Conn :: connection(),
Query :: iodata(),
Timeout :: default_timeout \| timeout(),
Params :: no_params \| [term()],
FilterMap :: no_filtermap_fun \| query_filtermap_fun(),
Result :: query_result().
```
which I am reluctant to do, since it is in fact incorrect as it suggests you can pass in two filtermap functions.

Reply:
Keep the good spec. It's more important than EDoc. I would rather drop EDoc if it's not good enough, but that can be done later. (What I do like about EDoc though is that you get the functions and types from the code for free.)

Can you post the link to the bug you opened for EDoc? I don't think it can be that hard to solve... (or is it?)

Reply:
Oh, sure, here you go: [ERL-849](https://bugs.erlang.org/browse/ERL-849). Not sure about how complicated it might be, have only glimpsed into the edoc code. `edoc_specs` seems to be the module to go for, I guess.

Reply:
Ok, [ERL-849](https://bugs.erlang.org/browse/ERL-849) has just been resolved as "won't fix"...